web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts

```diff
@@ -0,0 +1,210 @@
+    .transform(s => s.trim().split(/\s?\+\s?/))
+    .transform(putModifiersFirst(allowedModifiers))
+    .superRefine(validateKeyCodeAndModifiers(allowedModifiers))
+    .transform(adjustCasing)
+    .transform(s => s.join('+'));
+}
+
+function putModifiersFirst(
+  allowedModifiers: string[]
+): (tokens: string[]) => string[] {
+  return tokens =>
+    tokens.sort((a, b) => {
+      if (allowedModifiers.indexOf(a) === -1) {
+        return 1;
+      }
+      if (allowedModifiers.indexOf(b) === -1) {
+        return -1;
+      }
+      return allowedModifiers.indexOf(a) - allowedModifiers.indexOf(b);
+    });
+}
+
+/** Currently, works only for single characters.
+ * An accelerator string is compared to a string generated from event properties
+ * (so it should be case-insensitive too).
+ * But what is more important, this string is used also for tooltips.
+ * If we allow "CTRL+SHIFT+PAGEUP" then we cannot display it as "Ctrl+Shift+PageUp".
+ */
+function adjustCasing(tokens: string[]): string[] {
+  return tokens.map(token => {
+    if (token.length === 1) {
+      return token.toUpperCase();
+    }
+    return token;
+  });
+}
+
+function validateKeyCodeAndModifiers(
+  allowedModifiers: string[]
+): (tokens: string[], ctx: z.RefinementCtx) => void {
+  return (tokens, ctx) => {
+    // empty accelerator disables the shortcut
+    if (tokens.join('') === '') {
+      return z.NEVER;
+    }
+
+    const [keyCode, ...modifiers] = [...tokens].reverse();
+
+    const keyCodeUppercase =
+      keyCode.length === 1 // letters
+        ? keyCode.toUpperCase()
+        : keyCode;
+    if (!ALLOWED_KEY_CODES.includes(keyCodeUppercase)) {
+      ctx.addIssue(invalidKeyCodeIssue(keyCode));
+      return z.NEVER;
+    }
+
+    if (modifiers.length !== new Set(modifiers).size) {
+      ctx.addIssue(duplicateModifierIssue());
+      return z.NEVER;
+    }
+
+    const invalidModifiers = modifiers.filter(
+      modifier => !allowedModifiers.includes(modifier)
+    );
+    if (invalidModifiers.length) {
+      ctx.addIssue(invalidModifierIssue(invalidModifiers));
+      return z.NEVER;
+    }
+
+    if (!FUNCTION_KEYS.includes(keyCode) && !modifiers.length) {
+      ctx.addIssue(missingModifierIssue(keyCode));
+    }
+  };
+}
+
+/** Returns allowed modifiers for a given platform in the correct order. */
+function getSupportedModifiers(platform: Platform): string[] {
+  switch (platform) {
+    case 'win32':
+    case 'linux':
+      return ['Ctrl', 'Alt', 'Shift'];
+    case 'darwin':
+      return ['Cmd', 'Ctrl', 'Option', 'Shift'];
+  }
+}
+
+function generateRange(start: number, end: number): string[] {
+  return new Array(end - start + 1)
+    .fill(undefined)
+    .map((_, i) => (i + start).toString());
+}
+
+const FUNCTION_KEYS = generateRange(1, 24).map(key => `F${key}`);
+
+// subset of https://github.com/electron/electron/blob/49df19214ea3abaf0ad91adf3d374cba32abd521/docs/api/accelerator.md#available-key-codes
+const ALLOWED_KEY_CODES = [
```

Comment:
The grave accent (`) is advertised as an allowed key code but if you set a shortcut with it it doesn't seem to work.

Reply:
I think it is related to [dead key](https://en.wikipedia.org/wiki/Dead_key). On my keyboard it doesn't generate an event when it is pressed with Command. Probably we should remove it?

Reply:
Okay, so:

* Cmd+grave doesn't do anything because it's a macOS shortcut for cycling through open windows of the current app. 🤦
* grave is interpreted as IntlBacktick. https://github.com/gravitational/teleport/blob/7e75123f1fd6bac4d549eeeae8ba35e7614ff7e6/web/packages/teleterm/src/ui/services/keyboardShortcuts/getKeyName.ts#L78
* I have international layout on my MBP but I have Karabiner set up so that IntlBacktick actually sends regular grave to.
* Everything works fine after I removed that setting from Karabiner and bound next tab to Shift + grave.

Reply:
So, to sum up, we can leave it as it is?

Reply:
I switched to a clean profile in Karabiner to make sure it doesn't interfere with my findings here.

<details>
<summary>Reference of keyboard layouts</summary>

W3 writing system keys

![keyboard-codes-alphanum1](https://user-images.githubusercontent.com/27113/222691229-60b14cf2-d8a4-48d7-a2e3-2383dba952c9.svg)

[MacBook keyboard layouts](https://mwichary.medium.com/international-apple-keyboards-layouts-93437d7f9273)

![MacBook keyboard layouts](https://user-images.githubusercontent.com/27113/222691575-04853d64-c5ac-473c-83ce-2578508cc246.png)
</details>

I have U.S. International on my MBP. Let’s consider the key to the left of 1 (called top left) and to the left of Z (called bottom left).

* In Karabiner EventViewer
    * Top left sends non_us_backslash and inputs §
    * Bottom left sends grave_accent_and_tilde and inputs `
* In keydown handler:
    * Top left
        * Event: {code: “IntlBackslash”, key: “§”}
        * getKeyName `
        * accelerator `
    * Bottom left
        * Event: {code: “Backquote”, key: “`”}
        * getKeyName Backquote
        * accelerator Backquote

I think it might be an oversight in Tabby that IntlBackslash is mapped to \` (and I just found a bug report https://github.com/Eugeny/tabby/issues/5280). It feels that Backquote should be mapped to ` and IntlBackslash should be added to allowed key codes. IntlBackslash is not present in [Electron's list of available key codes for accelerators](https://github.com/electron/electron/blob/17ccb3c6ecca1b16fb3e10504ffe0135b4e1794c/docs/api/accelerator.md#available-key-codes) but that shouldn't matter as we're using browser standards here and not those of Electron as we're inside a keydown handler.

I checked it and after applying the below patch, I'm able to set shortcuts for both buttons.

<details>
<summary>Patch</summary>

```diff
diff --git a/web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts b/web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts
index 3d2df34141..a3618244af 100644
--- a/web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts
+++ b/web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts
@@ -166,6 +166,6 @@ const ALLOWED_KEY_CODES = [
   'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
   'Space', 'Tab', 'CapsLock', 'NumLock', 'ScrollLock', 'Backspace', 'Delete', 'Insert', 'Enter',
   'Up', 'Down', 'Left', 'Right',
-  'Home', 'End', 'PageUp', 'PageDown', 'Escape',
+  'Home', 'End', 'PageUp', 'PageDown', 'Escape', 'IntlBackslash',
   ',', '.', '/', '\\', '`', '-', '=', ';', "'", '[', ']',
 ];
diff --git a/web/packages/teleterm/src/ui/services/keyboardShortcuts/getKeyName.ts b/web/packages/teleterm/src/ui/services/keyboardShortcuts/getKeyName.ts
index 8f19c4356b..8835210e2f 100644
--- a/web/packages/teleterm/src/ui/services/keyboardShortcuts/getKeyName.ts
+++ b/web/packages/teleterm/src/ui/services/keyboardShortcuts/getKeyName.ts
@@ -75,7 +75,7 @@ const PHYSICAL_CODE_TO_PRINTABLE_CHARACTER = {
   Period: '.',
   Slash: '/',
   Backslash: '\\',
-  IntlBackslash: '`',
+  Backquote: '`',
   Minus: '-',
   Equal: '=',
   Semicolon: ';',
```
</details>

Reply:
It makes sense, done. 
Thanks for digging into this.